ocaml/sdk-gen/go/autogen/src/jsonrpc_client.go

```diff
@@ -0,0 +1,225 @@
+package xenapi
+
+import (
+	"bytes"
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"os"
+	"reflect"
+	"strings"
+	"time"
+)
+
+type Request struct {
+	JSONRPC string      `json:"jsonrpc"`
+	Method  string      `json:"method"`
+	Params  interface{} `json:"params,omitempty"`
+	ID      int         `json:"id"`
+}
+
+type ResponseError struct {
+	Code    int         `json:"code"`
+	Message string      `json:"message"`
+	Data    interface{} `json:"data,omitempty"`
+}
+
+type Response struct {
+	JSONRPC string         `json:"jsonrpc"`
+	Result  interface{}    `json:"result,omitempty"`
+	Error   *ResponseError `json:"error,omitempty"`
+	ID      int            `json:"id"`
+}
+
+func paramsParse(params ...interface{}) interface{} {
+	var finalParams interface{}
+	finalParams = params
+	if len(params) == 1 {
+		if params[0] != nil {
+			var typeOf reflect.Type
+			typeOf = reflect.TypeOf(params[0])
+			for typeOf != nil && typeOf.Kind() == reflect.Ptr {
+				typeOf = typeOf.Elem()
+			}
+			typeArr := []reflect.Kind{reflect.Struct, reflect.Array, reflect.Slice, reflect.Interface, reflect.Map}
+			if typeOf != nil {
+				for _, value := range typeArr {
+					if value == typeOf.Kind() {
+						finalParams = params[0]
+						break
+					}
+				}
+			}
+		}
+	}
+	return finalParams
+}
+
+type rpcClient struct {
+	endpoint   string
+	httpClient *http.Client
+	headers    map[string]string
+}
+
+func (client *rpcClient) newRequest(ctx context.Context, req interface{}) (*http.Request, error) {
+	dataByte, err := json.Marshal(req)
+	if err != nil {
+		return nil, fmt.Errorf("error marshaling request: %w", err)
+	}
+
+	request, err := http.NewRequestWithContext(ctx, http.MethodPost, client.endpoint, bytes.NewReader(dataByte))
+	if err != nil {
+		return nil, fmt.Errorf("error creating request: %w", err)
+	}
+
+	request.Header.Set("Content-Type", "application/json; charset=utf-8")
+	request.Header.Set("Accept", "application/json")
+	request.Header.Set("User-Agent", "XenAPI/" + APIVersionLatest.String())
+	for k, v := range client.headers {
+		request.Header.Set(k, v)
+	}
+
+	return request, nil
+}
+
+func convertUnhandledJSONData(jsonBytes []byte) []byte {
+	jsonString := string(jsonBytes)
+	jsonString = strings.ReplaceAll(jsonString, ":Infinity", ":\"+Inf\"")
```

Comment:
Does this replace Infinity/-Infinity with +Inf/-Inf in the response data? I thought that these were not valid Json and xapi was recently corrected to not use them (in fact we used to patch the C# Json library to understand them).

Reply:
Yes, we call `convertUnhandledJSONData` in `func (client *rpcClient) call(ctx context.Context, methodName string, params ...interface{}) (*Response, error)` :
```go
        var rpcResponse *Response
	body, err := io.ReadAll(httpResponse.Body)
	if err != nil {
		return nil, fmt.Errorf("call %v() on %v status code: %v. Could not read response body: %w", request.Method, httpRequest.URL.Redacted(), httpResponse.StatusCode, err)
	}
	body = convertUnhandledJSONData(body)
``` 
And when we deserialize float, we will call `deserializeFloat`:
```go
func deserializeFloat(context string, input interface{}) (value float64, err error) {
	_ = context
	if input == nil {
		return
	}
	strValue := fmt.Sprintf("%v", input)
	value, err = strconv.ParseFloat(strValue, 64)
	if err != nil {
		switch strValue {
		case "+Inf":
			return math.Inf(1), nil
		case "-Inf":
			return math.Inf(-1), nil
		case "NaN":
			return math.NaN(), nil
		}
	}
	return
}
```